pageserver/src/tenant/timeline/compaction.rs

```diff
@@ -251,9 +253,9 @@ impl Timeline {
 
             // Don't bother re-writing a layer if it is within the PITR window: it will age-out eventually
             // without incurring the I/O cost of a rewrite.
-            if layer_desc.get_lsn_range().end >= pitr_cutoff {
-                debug!(%layer, "Skipping rewrite of layer still in PITR window ({} >= {})",
-                    layer_desc.get_lsn_range().end, pitr_cutoff);
+            if layer_desc.get_lsn_range().end >= last_gc_cutoff {
+                debug!(%layer, "Skipping rewrite of layer still in GC window ({} >= {})",
+                    layer_desc.get_lsn_range().end, last_gc_cutoff);
```

Comment:
Is this not the wrong way around?

As in the layers that are "alive in the future" are within the cutoff. Or am I again thinking this the wrong way around?

Reply:
The idea is to rewrite only old layers: more recent layers are allowed to continue existing, because they'll likely get cleaned up anyway during GC.  This is to avoid doing a full rewrite of all the layers when we split a tenant: we instead tolerate some size bloat until the layers reach the GC cutoff, then if they _still_ exist, we interpret that as "this layer is probably pinned by some branch point, so rather than keep it forever in the unfiltered state, let's filter it"

Reply:
Aah ok so my point was correct but that is wanted behaviour. Ok makes sense.

How to stop me from thinking this is a bug in future :) You provided good rationale, however, this is racy. We compact (30/h) and refresh gc info (6/h) much more often than we do, however the `Timeline::latest_gc_cutoff_lsn` is advanced at once per hour. If last_gc_cutoff is `latest_gc_cutoff_lsn` this should be working as expected.

Idea for documenting the why so I don't report this as a bug when I go over it in editor:

```suggestion
            if layer_desc.get_lsn_range().end >= last_gc_cutoff {
                // intentionally only rewrite layers eligible for gc which must be
                // retained for branches
                debug!(%layer, "Skipping rewrite of layer still in GC window ({} >= {})",
                    layer_desc.get_lsn_range().end, last_gc_cutoff);
```

However ... This does make me think this is another case of possible gc and compaction problem? If we rewrite a layer which is doomed to be gc'd, removing it would be a panic?

I still haven't gotten to writing that test case.

Reply:
> I'll edit this soon upon confirming this does the read+write locking on Timeline::layers.

Yes, this exhibits the same pattern but the layer locking is now obscured under `Timeline::rewrite_layers` -- I think that is not great because it's not a free standing operation because of underlying implementation, I'd just prefer to see it inlined. I think this is more risky than current legacy compaction because we explicitly target the gc candidates.

The risk would be mitigated by using AND keeping alive the `Timeline::latest_gc_cutoff_lsn.read()` until done.

Reply:
so in 01504407e, I'm keeping that RcuReadGuard, but afaict that doesn't entirely exclude concurrent GC, it just prevents it entering.  So I've also made rewrite_layers safe to use even if some of the layers were already dropped, by trimming the layers to drop/rewrite _after_ taking the layer map write guard.

Ultimately it probably makes sense to just run not do compaction and GC at the same time, but let's not get into that in this PR

Reply:
Yeah I think this PR has a lot of extra stuff already.